### ISSUE TITLE
🚀 Feature: Allow dynamic reloading of secrets without requiring a restart

### ISSUE THREAD

**Issue by mrust1219** [backstage/backstage#26853]
### 🔖 Feature description

Currently, Backstage requires a full restart to load updated secrets (such as user/password) from Vault HashiCorp. Many organizations have password policies that require regular password rotations (every 60 or 90 days), and secrets like FID/Token passwords are refreshed dynamically in Vault.

### 🎤 Context

The current behavior of requiring a restart to load new secrets introduces operational inefficiencies and downtime. It would be beneficial to have a feature where Backstage can automatically load updated secrets from Vault without requiring a restart, reducing downtime and improving operational flexibility.

### ✌️ Possible Implementation

Implement functionality that allows Backstage to dynamically reload secrets (user/password) from Vault HashiCorp as soon as they are updated, without requiring a service restart.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by benjdlambert** [comment]
Yeah, we've spoken around this a little bit for database auth that might be rotated too.

It's a little tricky as some plugins might end up reading the config once on startup and then never reading the config again, so even if we're to update the config, it would still need to be rolled out on a plugin level.

I think that there are change events for the config, but that probably wont include any external secrets that get templated in, like environment variables for instance.

How do you load secrets in from vault right now?

**Comment by sanderaernouts** [comment]
Microsoft has a nice solution for this in .NET. They call it the [Options pattern](https://learn.microsoft.com/en-gb/dotnet/core/extensions/options). The idea is that you bind a specific piece of configuration to a specific type and have the DI framework inject a typed instance of the configuration instead. For example, if you use `IOptionsSnapshot<T>` in an ASP.NET Core controller method the backing configuration will be read on every new request due to the way the dependency is scoped in the DI framework (of .NET).

That makes it easier to deal with refreshing configuriont, but you still have to be aware of it and support it explicitly. With this pattern you have the same problems @benjdlambert mentioned, if someone reads directly from the configuration, uses `IOption<T>` instead of `IOptionsSnapshot<T>`, or injects `IOptionsSnapshot<T>` into a singleton instance the underlying configuration will never be refreshed. I guess it's always up to the plugin author to deal with the fact that configuration may change.

**Comment by tomislacker** [comment]
This should not be stale. There should still be a framework for reloading runtime information, secrets or otherwise, at runtime.
